### ISSUE TITLE
autofix corrupted data files

### ISSUE THREAD

**Issue by RoboMWM** [GriefPrevention/GriefPrevention#1068]
**As** a server admin
**I want** the plugin to automatically recreate corrupted data files
**Because** I'm clueless when a stacktrace appears in my console

Scenario: the playerdata file "dummyplayerdata" is corrupted
When I start up my server
And GP loads
And is unable to parse a playerdata file
Then it will log the file contents to the GP log
And delete the playerdata file

**Comment by nouish** [comment]
I think this ignores the actual problem; the _cause_ of data corruption/loss.

Can we instead work out a plan to make GP not prone to data loss and corruption?

The default, no administration data store, should use atomic writes. This would mean not overwriting existing data with partial or no data, but is hard to accomplish with the current pile-of-files, as these could end up out of sync.

My proposed idea is to use an embedded [transactional](https://www.sqlite.org/transactional.html) database, like SQLite, which is already provided by CraftBukkit, as GP's [application file format](https://www.sqlite.org/appfileformat.html). Done right, SQLite checks all the boxes, _and_ makes GP atomic, consistent, isolated and durable. For you, as a maintainer, this leads to less work down the line, not having to respond to issues about data losses. It even makes it easier for add-ons to associate their data with GP's core data, in a relational way.

[SQLite is the most widely deployed and used database engine](https://www.sqlite.org/mostdeployed.html), which talks for itself when it comes to future maintainability. Users may upload their database in issues to help pinpoint problems, from which it would be possible to make a minimal database to recreate the problem, that could be used in tests to protect against regressions.

Because SQLite is quite similiar to MySQL, this would make for a more robust MySQL implementation too, almost "for free".

Interested in contributing to this, but as this naturally breaks some APIs I hope we can come to some kind of conclusion first :)

**Comment by Jikoo** [comment]
I personally like the idea of swapping to SQLite as the default with potential support for other databases.

> It even makes it easier for add-ons to associate their data with GP's core data, in a relational way.

Addons should not be dependent on which storage implementation GP is using. I understand that setting up relational ties would make less active data management work for the addon, but then the addon also has to implement its own handling for each and every possible database type GP could support.

> Interested in contributing to this, but as this naturally breaks some APIs I hope we can come to some kind of conclusion first

Why would it break anything? The first step would be to write a SQLiteDataStore that only kicks in when the JDBC URL indicates a SQLite database. Breaking changes (like a more Future-oriented approach) could come later.

**Comment by nouish** [comment]
Massive thank you for your feedback, Jikoo 😄

> Addons should not be dependent on which storage implementation GP is using. I understand that setting up relational ties would make less active data management work for the addon, but then the addon also has to implement its own handling for each and every possible database type GP could support.

Very valid points. I should have clarified, but the idea I have in mind here is to allow GP-core to hold some metadata for players and claims. Something like `(addon_id INT, claim_id BIGINT, key_id INT, value BIGINT)` letting add-ons associate `long`s that can be whatever it wishes. Be it IDs in their own data set or what not.

> Why would it break anything? The first step would be to write a SQLiteDataStore that only kicks in when the JDBC URL indicates a SQLite database. Breaking changes (like a more Future-oriented approach) could come later.

I think it pointless to make the bowl of spaghetti that is data store logic any larger. More so when the plan for the future _IS_ to break this down and allow add-ons to decide the data store implementation. I would much rather make incremental improvements to the existing implementations, before adding code that would then have to be changed aswell. Generally, my view is that breaking down and changing the API *first* would be benefitial, as that then gives new, aswell as existing, implementations more freedoms. The example in mind here, is that a database implementation would *NOT* currently be able to autogenerate IDs, per https://github.com/TechFortress/GriefPrevention/pull/465. To me it makes little sense to assign an ID *BEFORE* an entity is persistent. My recommendation here is to split this into two separate events: one before the claim is registered, which is `Cancellable`, and one after, which is not, where the associated claim has indeed been assigned an ID.

Currently the MySQL implementation uses a separate table, `griefprevention_nextclaimid`, to decide the next ID to use. What would be a single action with `RETURN_AUTOGENERATED_KEYS` is here three (autocommit) transactions: delete next id, set next id, register claim. Sure this may be a non-issue if it was at least done as a single transaction, but would be nice to let the database implementations handle this automatically 😄

___

Looking forward to more discussion on this topic, as I think it paramount for GP not to ever lose data. We should keep in mind that (I believe) most users are young people who just want to play Minecraft with others, who are likely not taking as much time creating backups as they perhaps should be!

**Comment by nouish** [comment]
Forgot another relevant point here..!

> The first step would be to write a SQLiteDataStore that only kicks in when the JDBC URL indicates a SQLite database. Breaking changes (like a more Future-oriented approach) could come later.

I think it'd be more userfriendly to drop the JDBC URL from configuration (of course porting the old format for them!), and instead take a `Host` and `Name` value, where the former would be in format `host[:port]`, where the port is optional, and defaults to the MySQL port (seeing as only MySQL is supported as of now). The latter would be the name of the database to use.

JDBC is the API for database connectivity in Java. It's not very friendly for non-technical users, and is really an implementation detail. As for datasource properties, that you can add to the JDBC-string now; those could be provided as a `Properties` configuration.

Visual example:
```yml
GriefPrevention:
  Database:
    Type: mysql # [mysql, sqlite, ...]
    # SQLite / Other embedded
    File: database.db
    # MySQL / Other external
    Host: localhost:3306 # Since 3306 is the MySQL default, it could be omitted
    Name: minecraft
    UserName: <username>
    Password: <password>
    Properties:
      cachePrepStmts: true
      prepStmtCacheSize: 250
```

Instead of the current (except SQLite/embedded, which currently does not exist):

```yml
GriefPrevention:
  Database:
    URL: jdbc:mysql://localhost:3306/minecraft?cachePrepStmts=true&prepStmtCacheSize=250
    UserName: <username>
    Password: <password>
```

**Comment by Jikoo** [comment]
I think we're on the same page in terms of what we want out of a future state datastore.

> The idea I have in mind here is to allow GP-core to hold some metadata for players and claims. Something like (addon_id INT, claim_id BIGINT, key_id INT, value BIGINT) letting add-ons associate longs that can be whatever it wishes. Be it IDs in their own data set or what not.

I originally had written a large paragraph in reply to this, but I'm deleting it for this statement: I would like to offer addons a data storage solution but I think that's a large enough feature to warrant its own separate discussion. I would not include this in the first rewrite of the datastore because it adds a lot of complexity that could just as easily be a feature later.

> I think it pointless to make the bowl of spaghetti that is data store logic any larger. More so when the plan for the future IS to break this down and allow add-ons to decide the data store implementation. I would much rather make incremental improvements to the existing implementations, before adding code that would then have to be changed aswell. Generally, my view is that breaking down and changing the API first would be benefitial, as that then gives new, aswell as existing, implementations more freedoms.

The problem is that we know there are a lot of breaking changes in the pipe. While you're correct, it would be easiest on developers of GP core to make breaking changes as soon as possible, we're talking breaking most addons multiple times. Instead of a full breaking rewrite (which was making little headway) we've opted to gradually iterate towards future state. Breaking changes are slated for v19 I believe. If you want to go for it go nuts (though I'd wait for Robo's okay, I'm more of a consultant here), but be aware that it is unlikely to be pulled in the near future if it's breaking. See the [V20 discussion thread](https://github.com/TechFortress/GriefPrevention/discussions/63) for info on the roadmap.

> To me it makes little sense to assign an ID BEFORE an entity is persistent. My recommendation here is to split this into two separate events: one before the claim is registered, which is Cancellable, and one after, which is not, where the associated claim has indeed been assigned an ID.

I definitely agree with this, the way claim IDs are handled leaves a lot to be desired.

> only MySQL is supported as of now

I made the same assumption in the past. Pretty much any SQL implementation is supported provided the user has the acumen to install the JDBC driver, create a database, and set the JDBC URL properly. That's the reason (I believe, at least. hard to know for sure what was going through BigScary's head) the SQL usage is so basic and terrible.

**Comment by nouish** [comment]
> I originally had written a large paragraph in reply to this, but I'm deleting it for this statement: I would like to offer addons a data storage solution but I think that's a large enough feature to warrant its own separate discussion. I would not include this in the first rewrite of the datastore because it adds a lot of complexity that could just as easily be a feature later.

Understandable, and I agree. It would be senseless to offer this as an incomplete or not well thought out feature, when that would likely lead to breaking changes in the future, and those are bad enough, but more so when it's about persistent data.

> The problem is that we know there are a lot of breaking changes in the pipe. While you're correct, it would be easiest on developers of GP core to make breaking changes as soon as possible, we're talking breaking most addons multiple times. Instead of a full breaking rewrite (which was making little headway) we've opted to gradually iterate towards future state. Breaking changes are slated for v19 I believe. If you want to go for it go nuts (though I'd wait for Robo's okay, I'm more of a consultant here), but be aware that it is unlikely to be pulled in the near future if it's breaking. See the [V20 discussion thread](https://github.com/TechFortress/GriefPrevention/discussions/63) for info on the roadmap.

I think it fair to at least `@Deprecate` functionality first for a transition period.

That is precisely why I want to discuss changes first. Saves any of our time being wasted on fruitless PRs, while also ensuring the end result is well thought out, with as few breaking changes as possible.

> > only MySQL is supported as of now
> 
> I made the same assumption in the past. Pretty much any SQL implementation is supported provided the user has the acumen to install the JDBC driver, create a database, and set the JDBC URL properly. That's the reason (I believe, at least. hard to know for sure what was going through BigScary's head) the SQL usage is so basic and terrible.

Valid point, though I'd argue with the `Type` configuration being used internally to construct the JDBC URL, it would leave this unchanged but without the exposed internals. Not a fan of "supporting" other untested database engines, but hey, at least *this* wouldn't be a regression.

For what it's worth, because the SQL used is very basic, the only immediate dialect I can find that would fail (on SQLite) is this:

https://github.com/TechFortress/GriefPrevention/blob/c235bb0a00436ee9d14ea92a78ba15c6a1ddd555/src/main/java/me/ryanhamshire/GriefPrevention/DatabaseDataStore.java#L111-L114

Without being able to give it a go right now, and other than the above code throwing currently, it **may** in fact be SQLite compatible already. Of course this changes nothing from the POV of wanting proper support, using transactions, for both SQLite and MySQL (and more, in the future).

___

**Edit:** Decided to try GP with `Database: jdbc:sqlite:GPDB.db` to see if it would work, first removing the MySQL dialect SQL pointed out above. It did not work. Turns out, the following code can be reached *BEFORE* setting the SQL variables (that are not just `static` constants):
 https://github.com/TechFortress/GriefPrevention/blob/c235bb0a00436ee9d14ea92a78ba15c6a1ddd555/src/main/java/me/ryanhamshire/GriefPrevention/DatabaseDataStore.java#L121

https://github.com/TechFortress/GriefPrevention/blob/c235bb0a00436ee9d14ea92a78ba15c6a1ddd555/src/main/java/me/ryanhamshire/GriefPrevention/DatabaseDataStore.java#L142

Which meant the following code was trying to effectively do `Connection.prepareStatement(null)`, which then made me wonder why the MySQL implementation works in the first place. Turns out the MySQL driver throws an `SQLException` for this, whereas the SQLite driver throws an `NullPointerException`, which is not caught. Every new install of GP will have the following warning when this occurs for DB users:

```
[20:20:49 INFO]: [GriefPrevention] Unable to set next schema version to 3.  Details:
[20:20:49 INFO]: [GriefPrevention] SQL String can not be NULL
```

Tl;dr there is a small bug/warning caused by initializing the SQL variables AFTER a call that depends on these being set. Should probably be constants in the first place.

With those changes however, GP will indeed run with SQLite.
